test/parse.js

```diff
@@ -76,6 +76,88 @@ test('parse()', function (t) {
         st.end();
     });
 
+    t.test('decode dot keys correctly', function (st) {
+        st.deepEqual(
+            qs.parse('name%252Eobj.first=John&name%252Eobj.last=Doe', { allowDots: false, decodeDotInKeys: false }),
+            { 'name%2Eobj.first': 'John', 'name%2Eobj.last': 'Doe' },
+            'with allowDots false and decodeDotInKeys false'
+        );
+        st.deepEqual(
+            qs.parse('name%252Eobj.first=John&name%252Eobj.last=Doe', { allowDots: false, decodeDotInKeys: true }),
+            { 'name%2Eobj.first': 'John', 'name%2Eobj.last': 'Doe' },
+            'with allowDots false and decodeDotInKeys true'
+        );
+
+        st.deepEqual(
+            qs.parse('name%252Eobj.first=John&name%252Eobj.last=Doe', { allowDots: true, decodeDotInKeys: false }),
+            { 'name%2Eobj': { first: 'John', last: 'Doe' } },
+            'with allowDots true and decodeDotInKeys false'
+        );
+        st.deepEqual(
+            qs.parse('name%252Eobj.first=John&name%252Eobj.last=Doe', { allowDots: true, decodeDotInKeys: true }),
+            { 'name.obj': { first: 'John', last: 'Doe' } },
+            'with allowDots true and decodeDotInKeys true'
+        );
```

Comment:
these are all 4 of the combinations of these two options. the problem is that there's only 3 variations here.

We need to figure out if having 4 variations makes sense, and what that behavior would be - and if we can't come up with anything, then that 4th combination needs to throw.

Reply:
I think the second combination should throw here `with allowDots false and decodeDotInKeys true` because it ideally means decoding to invalid state like `{ name.obj.first: 'John', name.obj.last: 'Doe' }` which makes it very ambiguous. What do you think?

Reply:
the point of `allowDots` is to treat a dot in a key as a nested object marker. when false, `name.obj.first` would just be either literally `name.obj.first` or `name%2Eobj%2Efirst`, no? what about that is invalid?

Reply:
@ljharb, the problem is that round-tripping breaks here. Let's consider the following example: `qs.stringify({ 'name.obj': { first: 'John', last: 'Doe' } }, { allowDots: true, encodeDotInKeys: false })`. Now, the counter `options` for the string generated by this would be `{ allowDots: true, decodeDotInKeys: false }`, and the consumer might think that they would get the object they provided to `stringify`. However, the parser cannot distinguish between a key and a dot notation for key/value separation in `'name.obj.first=John&name.obj.last=Doe'`.

```javascript
qs.parse(qs.stringify({ 'name.obj': { first: 'John', last: 'Doe' } }, { allowDots: true, encodeDotInKeys: false }),{ allowDots: true, decodeDotInKeys: false } )
// { name: { obj: { first: 'John', last: 'Doe' } } }
```

So, I believe both combinations (one for stringify and one for parse) should be considered illegal, right? One could argue that if `stringify` already throws for the combination, then the string would never be generated, but we never know how the consumer tends to use the package. What do you think? 🤔

Reply:
In that case, should `encodeDotInKeys` *imply* `allowDots: true` when allowDots is not passed? it'd still throw when allowDots is false.

Reply:
Hm, I see why current behaviour works but is confusing. As a consumer I would expect `{ encodeDotInKeys: true }` to encode the keys. But currently we do not with this line `var encodedKey = allowDots && encodeDotInKeys ? key.replace(/\./g, '%2E') : key;`. I will fix it and push a change.